scanner/indexer/indexer.go

```diff
@@ -57,15 +74,151 @@ func NewIndexer(ctx context.Context) (*Indexer, error) {
+		Hash: manifestDigest,
+	}
+	zlog.Info(ctx).
+		Str("image_reference", imgRef.String()).
+		Int("layers_count", len(imgLayers)).
+		Msg("retrieving layers to populate container image manifest")
+	for _, layer := range imgLayers {
+		ccDigest, layerDigest, err := getLayerDigests(layer)
+		if err != nil {
+			return nil, fmt.Errorf("getting layer digests: %w", err)
+		}
+		// TODO Check for non-retriable errors (permission denied, etc.) to report properly.
+		layerReq, err := getLayerRequest(&httpClient, imgRef, layerDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting layer request URL and headers (digest: %q): %w",
+				layerDigest.String(), err)
+		}
+		layerReq.Header.Del("User-Agent")
+		layerReq.Header.Del("Range")
+		manifest.Layers = append(manifest.Layers, &claircore.Layer{
+			Hash:    ccDigest,
+			URI:     layerReq.URL.String(),
+			Headers: layerReq.Header,
+		})
+	}
+	return i.indexer.Index(ctx, manifest)
+}
+
+// getLayerDigests returns the clairclore and containerregistry digests for the layer.
+func getLayerDigests(layer v1.Layer) (ccd claircore.Digest, ld v1.Hash, err error) {
+	ld, err = layer.Digest()
+	if err != nil {
+		return ccd, ld, err
+	}
+	ccd, err = claircore.ParseDigest(ld.String())
+	return ccd, ld, err
+}
+
+// getLayerRequest sends a partial request to retrieve the layer from the
+// registry and return the request object containing relevant information to
+// populate a container image manifest.
+func getLayerRequest(httpClient *http.Client, imgRef name.Reference, layerDigest v1.Hash) (*http.Request, error) {
+	imgRepo := imgRef.Context()
+	registryURL := url.URL{
+		Scheme: imgRepo.Scheme(),
+		Host:   imgRepo.RegistryStr(),
+	}
+	imgPath := strings.TrimPrefix(imgRepo.RepositoryStr(), imgRepo.RegistryStr())
+	imgURL := path.Join("/", "v2", imgPath, "blobs", layerDigest.String())
+	u, err := registryURL.Parse(imgURL)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Range", "bytes=0-0")
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	utils.IgnoreError(res.Body.Close)
+	return res.Request, nil
+}
+
+// getContainerImageLayers fetches the image's manifest from the registry to get
+// a list of layers.
+func getContainerImageLayers(ctx context.Context, ref name.Reference, o options) ([]v1.Layer, error) {
+	// TODO Check for non-retriable errors (permission denied, etc.) to report properly.
+	desc, err := remote.Get(ref, remote.WithContext(ctx), remote.WithAuth(o.auth), remote.WithPlatform(o.platform))
+	if err != nil {
+		return nil, err
+	}
+	img, err := desc.Image()
+	if err != nil {
+		return nil, err
+	}
+	layers, err := img.Layers()
+	if err != nil {
+		return nil, err
+	}
+	return layers, nil
+}
+
+// parseContainerImageURL returns a image reference from an image URL.
+func parseContainerImageURL(imageURL string) (name.Reference, error) {
```

Comment:
For now, yes. It could be of interest elsewhere, but since I don't know right now, I will leave it private and keep the interfaces clean and obvious.